src/visitor.rs

```diff
@@ -47,6 +49,32 @@ fn is_extern_crate(item: &ast::Item) -> bool {
     }
 }
 
+/// Creates a string slice corresponding to the specified span.
+pub struct SnippetProvider {
+    /// A pointer to the content of the file we are formatting.
+    big_snippet: *const Rc<String>,
+    /// A position of the start of `big_snippet`, used as an offset.
+    start_pos: usize,
+}
+
+impl SnippetProvider {
+    pub fn span_to_snippet(&self, span: Span) -> Option<&str> {
+        let start_index = span.lo().to_usize().checked_sub(self.start_pos)?;
+        let end_index = span.hi().to_usize().checked_sub(self.start_pos)?;
+        unsafe { Some(&(*self.big_snippet)[start_index..end_index]) }
+    }
+
+    pub fn from_codemap(codemap: &CodeMap, span: Span) -> Self {
+        let filemap = codemap.lookup_char_pos(span.lo()).file;
+        let big_snippet = unsafe { mem::transmute(&filemap.src) };
```

Comment:
Rather than do this transmute, could we unwrap the option (or return Option<SnippetProvider>) and keep an Rc rather than `*const Rc<String>`? This looks pretty bad because we never increment the reference count on the Rc pointer, plus the transmute is relying on implementation details of the Rust compiler.